### PR TITLE
Implement user profile page

### PR DESCRIPTION
**Pull request by allenwq**
Depends on #441.

### CODE REVIEW

**Comment by lowjoel** [comment]
@allenwq rebase please then i'll review

File: app/controllers/user/omniauth_callbacks_controller.rb [hunk 1]
```diff
@@ -10,7 +10,12 @@ def facebook
   private
 
   def link_current_account_to_facebook
-    # TODO: Implement
+    auth = request.env['omniauth.auth']
+    if current_user && current_user.persisted? && current_user.link_to_ominiauth(auth)
+      redirect_to edit_user_profile_path, success: t('user.omniauth_callbacks.facebook.success')
+    else
+      redirect_to edit_user_profile_path, danger: t('user.omniauth_callbacks.facebook.failed')
```
Comment (lowjoel) [inline app/controllers/user/omniauth_callbacks_controller.rb]: probably you should give a reason.

File: app/controllers/user/omniauth_callbacks_controller.rb [hunk 2]
```diff
@@ -10,7 +10,12 @@ def facebook
   private
 
   def link_current_account_to_facebook
```
Comment (lowjoel) [inline app/controllers/user/omniauth_callbacks_controller.rb]: link_user_to_facebook

File: app/controllers/user/profiles_controller.rb [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+
+  def update
+    if current_user.update_attributes(profile_params)
+      redirect_to edit_user_profile_path, success: t('.success')
+    else
+      render 'edit'
+    end
+  end
+
+  def profile_params
```
Comment (lowjoel) [inline app/controllers/user/profiles_controller.rb]: private

File: app/models/concerns/user_omniauth_concern.rb [hunk 4]
```diff
@@ -36,4 +36,12 @@ def create_by_omniauth(auth)
       end
     end
   end
+
+  def link_to_ominiauth(auth)
+    identity = User::Identity.find_by(provider: auth.provider, uid: auth.uid)
```
Comment (lowjoel) [inline app/models/concerns/user_omniauth_concern.rb]: does find_or_initialize_by work?

File: config/locales/en/user/omniauth_callbacks.yml [hunk 5]
```diff
@@ -3,3 +3,5 @@ en:
     omniauth_callbacks:
       facebook:
         kind: 'Facebook'
+        success: 'Your account has been linked to your facebook successfully'
```
Comment (lowjoel) [inline config/locales/en/user/omniauth_callbacks.yml]: Your account has been successfully linked to Facebook.

File: app/controllers/user/omniauth_callbacks_controller.rb (hunk 1 above)
Comment (allenwq) [inline app/controllers/user/omniauth_callbacks_controller.rb]: Yah, I described the reason in the translation file.

File: app/controllers/user/profiles_controller.rb (hunk 3 above)
Comment (allenwq) [inline app/controllers/user/profiles_controller.rb]: oops, thanks !

File: app/models/concerns/user_omniauth_concern.rb (hunk 4 above)
Comment (allenwq) [inline app/models/concerns/user_omniauth_concern.rb]: I think it works, let me update the code.

File: app/views/user/profiles/edit.html.slim [hunk 6]
```diff
@@ -0,0 +1,13 @@
...
+
+= simple_form_for current_user, url: user_profile_path do |f|
+  = f.error_notification
+  = f.input :name, required: true
+
+  span
+    - unless current_user.identities.facebook.any?
+      div.facebook
+        = link_to user_omniauth_authorize_path(:facebook) do
+          = fa_icon 'facebook-square', text: t('.link_to_facebook'), class: 'fa-2x'
```
Comment (lowjoel) [inline app/views/user/profiles/edit.html.slim]: symbolise the fa_icon icon argument.

**Comment by lowjoel** [comment]
Does this depend on #475?

**Comment by allenwq** [comment]
This PR does not depend on #475

File: app/views/user/profiles/edit.html.slim (hunk 6 above)
Comment (allenwq) [inline app/views/user/profiles/edit.html.slim]: OK, but I think most of other places are using strings ? we want to use symbol afterwards ?

Comment (lowjoel) [inline app/views/user/profiles/edit.html.slim]: I'm not sure, hmm.

File: app/models/concerns/user_omniauth_concern.rb [hunk 7]
```diff
@@ -36,4 +36,15 @@ def create_by_omniauth(auth)
       end
     end
   end
+
+  # Link user with an omniauth provider.
+  #
+  # @param [OmniAuth::AuthHash] auth The data with omniauth info.
+  # @return [Bool] True if success, otherwise false.
+  def link_with_ominiauth(auth)
```
Comment (lowjoel) [inline app/models/concerns/user_omniauth_concern.rb]: should the facebook omniauth concern call this method?

Comment (allenwq) [inline app/models/concerns/user_omniauth_concern.rb]: no need, this function is general and for all the omniauth providers

Comment (lowjoel) [inline app/models/concerns/user_omniauth_concern.rb]: all the more the facebook omniauth concern should call this.

Comment (allenwq) [inline app/models/concerns/user_omniauth_concern.rb]: Sorry, I don't get you, how to call this ?

Comment (lowjoel) [inline app/models/concerns/user_omniauth_concern.rb]: in `new_with_session`, call `user.link_with_omniauth(info)` you will likely have to change the contract for link_with_omniauth not to save the record by default, instead the parent resource should save the change.

**Pull request merged**